File: Zeze/Transaction/Collections/PMap.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Immutable;

namespace Zeze.Transaction.Collections
{
    public abstract class PMap<K, V> : PCollection, IDictionary<K, V>
    {
        private readonly Func<ImmutableDictionary<K, V>, Log> _logFactory;
        protected ImmutableDictionary<K, V> map;

        public PMap(long logKey, Func<ImmutableDictionary<K, V>, Log> logFactory) : base(logKey)
        {
            this._logFactory = logFactory;
            map = ImmutableDictionary<K, V>.Empty;
        }

        public Log NewLog(ImmutableDictionary<K, V> value)
        {
            return _logFactory(value);
        }

        public abstract class LogV : Log
        {
            internal ImmutableDictionary<K, V> Value;
            protected LogV(Bean bean,ImmutableDictionary<K, V> value) : base(bean)
            {
                Value = value;
            }

            protected void Commit(PMap<K, V> variable)
            {
                variable.map = Value;
            }
        }

        protected ImmutableDictionary<K, V> Data
        {
            get
            {
                if (this.IsManaged)
                {
                    var txn = Transaction.Current;
                    if (txn == null)
                    {
                        return map;
                    }
                    txn.VerifyRecordAccessed(this, true);
                    return txn.GetLog(LogKey) is LogV log ? log.Value : map;
                }
                else
                {
                    return map;
                }
            }
        }

        [Obsolete("Don't use this, please use Keys2", true)]
        ICollection<K> IDictionary<K, V>.Keys => throw new NotImplementedException();
        [Obsolete("Don't use this, please use Values2", true)]
        ICollection<V> IDictionary<K, V>.Values => throw new NotImplementedException();

        public IEnumerable<K> Keys => Data.Keys;

        public IEnumerable<V> Values => Data.Values;

        public int Count => Data.Count;

        public override string ToString()
        {
            return $"PMap{Data}";
        }
        public bool IsReadOnly => false;

        public abstract V this[K key] { get; set; }
        public abstract void Add(K key, V value);
        public abstract void Add(KeyValuePair<K, V> item);
        public abstract void AddRange(IEnumerable<KeyValuePair<K, V>> pairs);
        public abstract void SetItem(K key, V value);
        public abstract void SetItems(IEnumerable<KeyValuePair<K, V>> items);
        public abstract void Clear();
        public abstract bool Remove(K key);
        public abstract bool Remove(KeyValuePair<K, V> item);

        public void CopyTo(KeyValuePair<K, V>[] array, int arrayIndex)
        {
            int index = arrayIndex;
            foreach (var e in Data)
            {
                array[index++] = e;
            }
        }

        public bool Contains(KeyValuePair<K, V> item)
        {
            return Data.Contains(item);
        }

        public bool ContainsKey(K key)
        {
            return Data.ContainsKey(key);
        }

        public bool TryGetValue(K key, out V value)
        {
            return Data.TryGetValue(key, out value);

        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return Data.GetEnumerator();
        }

        IEnumerator<KeyValuePair<K, V>> IEnumerable<KeyValuePair<K, V>>.GetEnumerator()
        {
            return Data.GetEnumerator();
        }

        public ImmutableDictionary<K, V>.Enumerator GetEnumerator()
        {
            return Data.GetEnumerator();
        }
    }

    public class PMapReadOnly<K, V, P> : IReadOnlyDictionary<K, V> where P : V
    {
        private readonly PMap<K, P> _origin;

        public PMapReadOnly(PMap<K, P> origin)
        {
            _origin = origin;
        }

        public V this[K key] => _origin[key];

        public IEnumerable<K> Keys => _origin.Keys;

        public IEnumerable<V> Values => (IEnumerable<V>)_origin.Values;

        public int Count => _origin.Count;

        public bool ContainsKey(K key) => _origin.ContainsKey(key);

        public IEnumerator<KeyValuePair<K, V>> GetEnumerator()
        {
            foreach (var e in _origin)
            {
                yield return new KeyValuePair<K, V>(e.Key, e.Value);
            }
        }

        IEnumerator IEnumerable.GetEnumerator() => ((IEnumerable)_origin).GetEnumerator();

        public bool TryGetValue(K key, out V value)
        {
            if (_origin.TryGetValue(key, out var cur))
            {
                value = cur;
                return true;
            }
            value = default;
            return false;
        }
    }
}

File: ZezeJava/src/Zeze/Transaction/Collections/PMap.java
package Zeze.Transaction.Collections;

import Zeze.Transaction.*;
import java.util.*;
import java.util.stream.Collectors;

import org.pcollections.Empty;

public abstract class PMap<K, V> extends PCollection implements Map<K, V> {
	private final LogFactory<org.pcollections.PMap<K, V>> _logFactory;
	protected org.pcollections.PMap<K, V> map;

	public PMap(long logKey, LogFactory<org.pcollections.PMap<K, V>> logFactory) {
		super(logKey);
		this._logFactory = logFactory;
		map = Empty.map();
	}

	public final Log NewLog(org.pcollections.PMap<K, V> value) {
		return _logFactory.create(value);
	}

	public abstract static class LogV<K, V> extends Log {
		public org.pcollections.PMap<K, V> Value;
		protected LogV(Bean bean, org.pcollections.PMap<K, V> value) {
			super(bean);
			Value = value;
		}

		public final void Commit(PMap<K, V> var) {
			var.map = Value;
		}
	}

	protected final org.pcollections.PMap<K, V> getData() {
		if (this.isManaged()) {
			var txn = Transaction.getCurrent();
			if (txn == null) {
				return map;
			}
			txn.VerifyRecordAccessed(this, true);
			var log = txn.GetLog(LogKey);
			@SuppressWarnings("unchecked")
			var oldv = null != log ? ((LogV<K, V>)log).Value : map;
			return oldv;
		}
		else {
			return map;
		}
	}

	public final int size() {
		return getData().size();
	}

	@Override
	public String toString() {
		return getData().entrySet().stream().map(e -> e.getKey().toString() + ":" + e.getValue().toString()).collect(Collectors.joining(",", "{", "}"));
	}

	public final boolean isReadOnly() {
		return false;
	}

	public abstract V put(K key, V value);
	public abstract void putAll(Map<? extends K, ? extends V> m);
	public abstract void clear();
	public abstract V remove(Object key);
	public abstract boolean remove(Map.Entry<K, V> item);

	public final void copyTo(Map.Entry<K, V>[] array, int arrayIndex) {
		int index = arrayIndex;
		for (var e : getData().entrySet()) {
			array[index++] = e;
		}
	}

	public V get(Object key) {
		return getData().get(key);
	}

	public final boolean containsValue(Object v) {
		return getData().containsValue(v);
	}

	public final boolean containsKey(Object key) {
		return getData().containsKey(key);
	}

    public Set<K> keySet() {
        return Collections.unmodifiableSet(getData().keySet());
    }

    public Collection<V> values() {
        return Collections.unmodifiableCollection(getData().values());
    }

    public Set<Map.Entry<K, V>> entrySet() {
        return Collections.unmodifiableSet(getData().entrySet());
    }

	@Override
	public boolean isEmpty() {
		return getData().isEmpty();
	}
}